### PR TITLE
[freerdp] move glib from dependencies to feature

### PR DESCRIPTION
**Pull request by KangLin**
- Move glib to feature
- Disable glib in android

### CODE REVIEW

File: ports/freerdp/vcpkg.json [hunk 1]
```diff
@@ -1,17 +1,13 @@
 {
   "name": "freerdp",
-  "version": "3.0.0",
+  "version": "3.3.0",
   "port-version": 1,
```
Comment (Cheney-W) [inline ports/freerdp/vcpkg.json:4]:
```suggestion
```

File: versions/f-/freerdp.json [hunk 2]
```diff
@@ -1,5 +1,15 @@
 {
   "versions": [
+    {
+      "git-tree": "ca523cf88256f0fa33f8359ec8e3cfcc4fd18122",
+      "version": "3.3.0",
+      "port-version": 1
+    },
```
Comment (Cheney-W) [inline versions/f-/freerdp.json:7]:
```suggestion
```

**Review by Cheney-W — CHANGES_REQUESTED** [review]

**Comment by Cheney-W** [comment]
1. Please remove the `port-version` item when you update the` version` item.
2. Please make sure that you only add one new record to the `versions/f-/freerdp.json` file in your PR.

**Comment by Cheney-W** [comment]
image
I think you should update the version item.

**Comment by dg0yt** [comment]
If you want a new version, you must also update portfile.

If you want to change features, you must also update the portfile.

File: ports/freerdp/vcpkg.json [hunk 3]
```diff
@@ -1,17 +1,12 @@
 {
   "name": "freerdp",
-  "version": "3.0.0",
-  "port-version": 1,
+  "version": "3.3.0",
```
Comment (dg0yt) [inline ports/freerdp/vcpkg.json:3]: I'm puzzled. How can the portfile download the new tarball without having the new SHA512?

https://dev.azure.com/vcpkg/public/_build/results?buildId=99922&view=logs&j=f79cfdd7-47a8-597f-8f57-dc3e21a8f2ad&t=da63cf11-1f12-503b-6d64-3b2fb713c172:
~~~
Installing 84/85 freerdp[core,ffmpeg,server,urbdrc,wayland,winpr-tools,x11]:x64-linux@3.3.0#1...
Building freerdp[core,ffmpeg,server,urbdrc,wayland,winpr-tools,x11]:x64-linux@3.3.0#1...
-- Downloading https://github.com/FreeRDP/FreeRDP/archive/3.3.0.tar.gz -> FreeRDP-FreeRDP-3.3.0.tar.gz...
-- Extracting source /mnt/vcpkg-ci/downloads/FreeRDP-FreeRDP-3.3.0.tar.gz
~~~

File: versions/f-/freerdp.json [hunk 4]
```diff
@@ -1,5 +1,15 @@
...
+    {
+      "git-tree": "ca523cf88256f0fa33f8359ec8e3cfcc4fd18122",
+      "version": "3.3.0",
+      "port-version": 0
+    },
+    {
+      "git-tree": "93eaeec0708e2da1d98ce3dacdfd16e24aeca98c",
+      "version": "3.0.0",
+      "port-version": 2
+    },
```
Comment (dg0yt) [inline versions/f-/freerdp.json:12]:
```suggestion
```

File: ports/freerdp/vcpkg.json (hunk 3 above)
Comment (dg0yt) [reply ports/freerdp/vcpkg.json:3]: NOT RESOLVED. YOU MUST UPDATE THE PORTFILE.
~~~
$ sha512sum 3.*.0.tar.gz 
6c9061674716ca8c83a3913222db4002d893d751b0072a8af10013e09462a9cc847689dc874e30c499ae0d5be73c464f610057744c771fcd678bc43185d0f923  3.0.0.tar.gz
6779d5d4098e0abf65c3c2a2321644a9c78bf7fb3cf0a692be7c88c7f04040c2fb54d63029a2771ead0453f3af17b64dd44136d387eb4a2a4586ccfb72a37c08  3.3.0.tar.gz
~~~

**Comment by dg0yt** [comment]
I'm puzzled. How can the portfile download the new tarball without having the new SHA512?

https://dev.azure.com/vcpkg/public/_build/results?buildId=99922&view=logs&j=f79cfdd7-47a8-597f-8f57-dc3e21a8f2ad&t=da63cf11-1f12-503b-6d64-3b2fb713c172:
~~~
Installing 84/85 freerdp[core,ffmpeg,server,urbdrc,wayland,winpr-tools,x11]:x64-linux@3.3.0#1...
Building freerdp[core,ffmpeg,server,urbdrc,wayland,winpr-tools,x11]:x64-linux@3.3.0#1...
-- Downloading https://github.com/FreeRDP/FreeRDP/archive/3.3.0.tar.gz -> FreeRDP-FreeRDP-3.3.0.tar.gz...
-- Extracting source /mnt/vcpkg-ci/downloads/FreeRDP-FreeRDP-3.3.0.tar.gz
~~~

**Comment by KangLin** [comment]
> I'm puzzled. How can the portfile download the new tarball without having the new SHA512?
> 
> https://dev.azure.com/vcpkg/public/_build/results?buildId=99922&view=logs&j=f79cfdd7-47a8-597f-8f57-dc3e21a8f2ad&t=da63cf11-1f12-503b-6d64-3b2fb713c172:
> 
> ```
> Installing 84/85 freerdp[core,ffmpeg,server,urbdrc,wayland,winpr-tools,x11]:x64-linux@3.3.0#1...
> Building freerdp[core,ffmpeg,server,urbdrc,wayland,winpr-tools,x11]:x64-linux@3.3.0#1...
> -- Downloading https://github.com/FreeRDP/FreeRDP/archive/3.3.0.tar.gz -> FreeRDP-FreeRDP-3.3.0.tar.gz...
> -- Extracting source /mnt/vcpkg-ci/downloads/FreeRDP-FreeRDP-3.3.0.tar.gz
> ```

This is the previous. Now it has passed the CI.

**Comment by dg0yt** [comment]
> This is the previous. Now it has passed the CI.

And we are puzzled as much as before *because* it has passed the CI. That download still doesn't have the SHA512 which is in the portfile.

**Comment by Neumann-A** [comment]
> And we are puzzled as much as before because it has passed the CI. That download still doesn't have the SHA512 which is in the portfile.

It doesn't. The asset cache just checks if the SHA512 is cached and pulls it independent of the file name. So CI never ran with the new tar file.

**Comment by dg0yt** [comment]
> So CI never ran with the new tar file.

Yeah, that's the right pointer to the effect of asset caching here.

Probably caused by contributors which didn't build the port with default (filename-based) asset caching.

However this means that CI doesn't catch a major problem which official reviewers don't catch early. 

And this means CI should disable reading from asset cache for ports with modified sources. Which CI still doesn't identify. (ABI hashes include port changes and changes of dependencies.)

**Comment by Cheney-W** [comment]
@KangLin 
After you've made changes to the `version` and `glib feature` in the `vcpkg.json` file, please install this port locally. (You'll also need to update the `SHA512` value in the `portfile.cmake` and modify some patches that aren't applicable to the new source.)

**Comment by KangLin** [comment]
move glib from dependencies to feature and disable glib in android

**Comment by dg0yt** [comment]
Again, you must edit the portfile. You did only the easy 20% of the work.

**Comment by KangLin** [comment]
> Again, you must edit the portfile. You did only the easy 20% of the work.
@dg0yt 
I've tested it successfully locally. The contents of the portfile have nothing to do with glib. What needs to be changed?

**Comment by dg0yt** [comment]
First you have to update the SHA512 of the download. Get the tarball manually.
Second you will probably have to update the patches.
Last, you will have to couple the feature to some options to cmake. If the feature isn't active, freerdp must not use glib, even when it is installed.

I would like to note the freerdp 3.0.0 and glib build successfully for android arm64 in vcpkg CI. The feature doesn't fix anything.

**Comment by KangLin** [comment]
> First you have to update the SHA512 of the download. Get the tarball manually. 

Don't update version. so the SHA512 is same.

> Second you will probably have to update the patches. Last, you will have to couple the feature to some options to cmake. If the feature isn't active, freerdp must not use glib, even when it is installed.

The patch is do with glib. the freerdp automation glib.

> 
> I would like to note that freerdp 3.0.0 and glib build successfully for android arm64 in vcpkg CI. The feature doesn't fix anything.

Only fix the compile error in android.

**Comment by dg0yt** [comment]
> The patch is do with glib. the freerdp automation glib.

Unclear.

> Only fix the compile error in android.

Which error do you refer to? There is no indication of an error. Actually, it is even enforced that freerdp must build in CI:
https://github.com/microsoft/vcpkg/blob/2ac6ba173be85207a7bac4f0f02f5486f9fd4c0f/scripts/ci.baseline.txt#L1276-L1284

**Comment by KangLin** [comment]
https://github.com/KangLin/RabbitRemoteControl/actions/runs/8092287866/job/22112761642#step:15:65

-- Found CMake 3.28.3
-- Android toolchain file within NDK detected: /usr/local/lib/android/sdk/ndk/25.2.9519653/build/cmake/android.toolchain.cmake
-- VCPKG-Info: CMAKE_MAP_IMPORTED_CONFIG_MINSIZEREL set to MinSizeRel;Release;None;
-- VCPKG-Info: CMAKE_MAP_IMPORTED_CONFIG_RELWITHDEBINFO set to RelWithDebInfo;Release;None;
-- Running vcpkg install
freerdp is only supported on '!(arm | uwp)', which does not match arm64-android. This usually means that there are known build failures, or runtime problems, when building other platforms. To ignore this and attempt to build freerdp anyway, rerun vcpkg with `--allow-unsupported`.
-- Running vcpkg install - failed
CMake Error at /home/runner/work/RabbitRemoteControl/vcpkg/scripts/buildsystems/vcpkg.cmake:899 (message):
Error:   vcpkg install failed.  See logs for more information:
  /home/runner/work/RabbitRemoteControl/RabbitRemoteControl/build/vcpkg-manifest-install.log
Call Stack (most recent call first):
  .cache/tools/qt/Qt/6.6.2/android_arm64_v8a/lib/cmake/Qt6/qt.toolchain.cmake:80 (include)
  /usr/local/share/cmake-3.28/Modules/CMakeDetermineSystem.cmake:170 (include)
  CMakeLists.txt:6 (project)

**Comment by dg0yt** [comment]
> `freerdp is only supported on '!(arm | uwp)', which does not match arm64-android.`

You are not building the latest version of the port. This restriction was removed with 2.10.0 in #32416, August 2023.

**Comment by KangLin** [comment]
The vcpkg VCPKGGITCOMMITID is 2ac6ba173be85207a7bac4f0f02f5486f9fd4c0f
It is up to date.

https://github.com/KangLin/RabbitRemoteControl/blob/develop/.github/workflows/android.yml

**Comment by KangLin** [comment]
On Windows, it gives the following error. so that I am think move glib to feature. disable glib in android on windows.

D:\source\vcpkg>vcpkg install freerdp:x64-android
Computing installation plan...
The following packages will be built and installed:
    freerdp:x64-android@3.0.0#1
  * glib:x64-android@2.78.4
Additional packages (*) will be modified to complete this operation.
Detecting compiler hash for triplet x64-android...
Restored 0 package(s) from C:\Users\M\AppData\Local\vcpkg\archives in 216 us. Use --debug to see more details.
Installing 1/2 glib:x64-android@2.78.4...
Building glib:x64-android@2.78.4...
-- Using cached glib-2.78.4.tar.xz.
-- Cleaning sources at D:/source/vcpkg/buildtrees/glib/src/glib-2-ad6932593c.clean. Use --editable to skip cleaning for the packages you specify.
-- Extracting source D:/source/vcpkg/downloads/glib-2.78.4.tar.xz
-- Applying patch use-libiconv-on-windows.patch
-- Applying patch libintl.patch
-- Using source at D:/source/vcpkg/buildtrees/glib/src/glib-2-ad6932593c.clean
-- Getting CMake variables for x64-android
-- Configuring x64-android-dbg
-- Using cached msys2-mingw-w64-x86_64-pkgconf-1~2.1.0-1-any.pkg.tar.zst.
-- Using cached msys2-msys2-runtime-3.4.10-4-x86_64.pkg.tar.zst.
-- Using msys root at D:/source/vcpkg/downloads/tools/msys2/fdbea3694fb5c0d4
-- Configuring x64-android-dbg done
-- Configuring x64-android-rel
-- Configuring x64-android-rel done
-- Package x64-android-dbg
CMake Error at scripts/cmake/vcpkg_execute_required_process.cmake:112 (message):
    Command failed: D:/source/vcpkg/downloads/tools/ninja/1.10.2-windows/ninja.exe install -v
    Working Directory: D:/source/vcpkg/buildtrees/glib/x64-android-dbg
    Error code: 1
    See logs for more information:
      D:\source\vcpkg\buildtrees\glib\package-x64-android-dbg-out.log
      D:\source\vcpkg\buildtrees\glib\package-x64-android-dbg-err.log

Call Stack (most recent call first):
  scripts/cmake/vcpkg_install_meson.cmake:33 (vcpkg_execute_required_process)
  ports/glib/portfile.cmake:56 (vcpkg_install_meson)
  scripts/ports.cmake:172 (include)


error: building glib:x64-android failed with: BUILD_FAILED
Elapsed time to handle glib:x64-android: 1.8 min
Please ensure you're using the latest port files with `git pull` and `vcpkg update`.
Then check for known issues at:
  https://github.com/microsoft/vcpkg/issues?q=is%3Aissue+is%3Aopen+in%3Atitle+glib
You can submit a new issue at:
  https://github.com/microsoft/vcpkg/issues/new?title=[glib]+Build+error+on+x64-android&body=Copy+issue+body+from+D%3A%2Fsource%2Fvcpkg%2Finstalled%2Fvcpkg%2Fissue_body.md

**Comment by dg0yt** [comment]
Do you know whether freerdp builds at all without glib on android? You might have saved time by a proper error report instead of jumping into the conclusion that freerdp needs a fix. 

So the issue is: glib fails to build for android targets on windows hosts.

The answer is that building for android on windows hosts is not tested in CI and known to have problems. Many ports, including glib, need msys2 to build on windows. 

I suggest that you close this PR, and that you properly report the glib issue, with the right log files. See https://github.com/microsoft/vcpkg/issues/30604 for how to report an issue.

**Comment by dg0yt** [comment]
[install-x64-android-dbg-out.log](https://github.com/microsoft/vcpkg/files/14444843/install-x64-android-dbg-out.log)

Try NDK r25.

**Comment by KangLin** [comment]
> Do you know whether freerdp builds at all without glib on android? You might have saved time by a proper error report instead of jumping into the conclusion that freerdp needs a fix.

Yes. I am sure to build freerdp without glib on android. The PR is passed. I am will test it in my program. so  I reserve my opinion.

**Comment by dg0yt** [comment]
> The PR is passed.

It passed, but it is not acceptable because it causes installation order problems. That's why you must update the portfile: You must actively prevent the use of glib if the feature is not active. Otherwise reinstalling a cached artifact might introduce an undeclared binary dependency on glib.

**Comment by dg0yt** [comment]
What I can confirm is that `cmake/FindGlib.cmake` was remove in 3.0.0-rc0.
We should check whether they removed that dependency completely.

**Comment by dg0yt** [comment]
More resources:
https://github.com/FreeRDP/FreeRDP/discussions/7655
https://github.com/FreeRDP/FreeRDP/wiki/Compilation

**Comment by KangLin** [comment]
@dg0yt 
in macos. the follow error:

```
-- Configuring done (22.5s)
CMake Error at build/vcpkg_installed/x64-osx/share/WinPR/WinPRTargets.cmake:61 (set_target_properties):
-- Generating done (0.3s)
Error:   The link interface of target "winpr" contains:

    Threads::Threads

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.
```

https://github.com/KangLin/RabbitRemoteControl/actions/runs/8105937330/job/22155076713#step:14:432

@akallabeth I think that in order to solve the problem fundamentally, it is still necessary to use vcpkg officially.

**Comment by dg0yt** [comment]
Probably a missing `find_dependency` (upstream bug, but should be patched).

**Comment by dg0yt** [comment]
So the cmake config needs another `find_dependency`.

**Comment by dg0yt** [comment]
> I think to solve the problem fundamentally, it is recommended that the official freerdp use vcpkg.

I don't think so. 
But vcpkg users must report errors. The vcpkg port was initiated by Windows users. It was a lot of work to reach the current state which really isn't bad.

**Comment by KangLin** [comment]
I intend to contribute directly to freerdp.
Can avoid maintaining two places of code

**Comment by akallabeth** [comment]
I don´t get your issues here, there seems to be too many patches (include paths/...) on top of upstream.

lets take `Threads::Threads`: the only direct use of that is in `SDL` client, in `WinPR` it is not used at all.

we´re building these in shared/static versions on quite a lot of platforms.

so, stuff like:
`"supports": "!uwp & !(arm & windows)",` - so where do these exceptions come from?

**Comment by dg0yt** [comment]
Do `gitk ports/freerdp` to see where we are coming from. It was a long way from somebody's works-for-me to mostly-works. And there are only you very few contributors doing thorough research to reduce patching. (Like: checking if `glib` is still required.)

`Threads::Threads` may come as a transitive dependency. The vcpkg CI configuration usually uncover a number hidden problems a) with transitive usage requirements for static linkage and b) with uncontrolled ("automatic") dependencies. And that's the key reason of most build system patching.

**Comment by akallabeth** [comment]
> Do `gitk ports/freerdp` to see where we are coming from. It was a long way from somebody's works-for-me to mostly-works. And there are only you very few contributors doing thorough research to reduce patching. (Like: checking if `glib` is still required.)
> 

well, I got that impression.

> `Threads::Threads` may come as a transitive dependency. The vcpkg CI configuration usually uncover a number hidden problems a) with transitive usage requirements for static linkage and b) with uncontrolled ("automatic") dependencies. And that's the key reason of most build system patching.

If I get a dependency from some library I link to then it is that library that is responsible for resolving that dependency.
(`pkg-config` or `CMake` files should list that in the dependencies)

yes, static linking is another beast, but setting up proper search paths for dependencies (and their dependencies) did work for our builds all the time on all platforms.

**Comment by dg0yt** [comment]
Updating to 3.3.0: https://github.com/microsoft/vcpkg/pull/37081

**Comment by BillyONeal** [comment]
Is this PR obsoleted by https://github.com/microsoft/vcpkg/pull/37081/ ?

**Comment by KangLin** [comment]
@BillyONeal  See: https://github.com/FreeRDP/FreeRDP/pull/9931

**Comment by KangLin** [comment]
@BillyONeal in https://github.com/FreeRDP/FreeRDP/pull/9931 , Now Freerdp may use VCPKG's manifest mode.

But how should the FreeRDP be done to integrate into vcpkg?
For example:
I have a [RabbitRemoteControl](https://github.com/KangLin/RabbitRemoteControl) program that depend on FreeRDP. 
The [vcpkg.json](https://github.com/KangLin/RabbitRemoteControl/blob/master/vcpkg.json) file.

**Comment by dg0yt** [comment]
> Is this PR obsoleted by #37081 ?

Yes.  #37081 was the result of reviewing  #37015.

> Now Freerdp may use VCPKG's manifest mode.
> But how should the FreeRDP be done to integrate into vcpkg?

What kind of problem do you need to resolve?
The manifest in upstream's FreeRDP is completely independent from the port's manifest in vcpkg. And only the latter is used when users consume `freerdp` in manifest mode (using the official vcpkg registry.) 

BTW you should better model your app's feature as:
~~~json
  "features": {
    "freerdp": {
      "description": "freerdp",
      "supports": "windows | linux"
      "dependencies": [
        "freerdp"
      ]
    },
~~~
assuming freerdp is disabled where it is not available. (And `windows` includes  `mingw`.)

**Comment by KangLin** [comment]
> What kind of problem do you need to resolve? The manifest in upstream's FreeRDP is completely independent from the port's manifest in vcpkg. And only the latter is used when users consume `freerdp` in manifest mode (using the official vcpkg registry.)

@dg0yt If this is the case, the freerdp official needs to maintain two codes. This is unacceptable. Is there a way to maintain just one place?

There are too many patches for this port, this should be resolved upstream.

**Comment by KangLin** [comment]
> What kind of problem do you need to resolve? The manifest in upstream's FreeRDP is completely independent from the port's manifest in vcpkg. And only the latter is used when users consume `freerdp` in manifest mode (using the official vcpkg registry.)

@dg0yt If this is the case, the freerdp official needs to maintain two codes. This is unacceptable. Is there a way to maintain just one place?

There are too many patches for this port, this should be resolved upstream.

**Comment by dg0yt** [comment]
> If this is the case, the freerdp official needs to maintain two codes. This is unacceptable. Is there a way to maintain just one place?

- The port isn't maintained by freerdp, but here by the community.
- A port is more than just a manifest. That's why it comes with different dependencies.

> There are too many patches for this port, this should be resolved upstream.

Feel free to help with that. I just pushed three of them. But for packages of this complexity, I wouldn't expect the count to go to zero. Sometimes a local patch is appopriate and still not a good fit for upstream integration.

**Pull request closed without merging**